File: app/src/main/java/org/mesonet/app/forecast/SemiDayForecastDataController.java
package org.mesonet.app.forecast;

import android.content.Context;
import android.content.res.Configuration;

import org.mesonet.app.R;
import org.mesonet.app.dependencyinjection.PerFragment;
import org.mesonet.app.formulas.DefaultUnits;
import org.mesonet.app.formulas.UnitConverter;
import org.mesonet.app.userdata.Preferences;

import java.util.Locale;
import java.util.Observable;

import javax.inject.Inject;

import static org.mesonet.app.userdata.Preferences.UnitPreference.kImperial;
import static org.mesonet.app.userdata.Preferences.UnitPreference.kMetric;


public class SemiDayForecastDataController extends Observable implements ForecastData
{
    Preferences mPreferences;

    UnitConverter mUnitConverter;

    ForecastModel mForecastModel;

    Context mContext;



    public SemiDayForecastDataController(Context inContext, Preferences inPreferences, UnitConverter inUnitConverter, ForecastModel inForecastModel)
    {
        mPreferences = inPreferences;
        mUnitConverter = inUnitConverter;
        mForecastModel = inForecastModel;
        mContext = inContext;
    }

    public void SetData(ForecastModel inForecast)
    {
        mForecastModel = inForecast;

        setChanged();
        notifyObservers();
    }

    @Override
    public String GetTime() {
        String result = mForecastModel.mTime;

        if((mContext != null && mContext.getResources().getBoolean(R.bool.forceWrapForecasts)) && !result.contains(" "))
            result += "\n";

        return result;
    }

    @Override
    public String GetIconUrl() {
        return mForecastModel.mIconUrl.replace("http://www.nws.noaa.gov/weather/images/fcicons", "http://www.mesonet.org/images/fcicons-android").replace(".jpg", "@4x.png");
    }

    @Override
    public String GetStatus() {
        return mForecastModel.mStatus + "\n";
    }

    @Override
    public String GetHighOrLow() {
        return mForecastModel.mHighOrLow.name();
    }

    @Override
    public String GetTemp() {
        UnitConverter.TempUnits tempUnits = UnitConverter.TempUnits.kCelsius;

        if(mPreferences == null || mPreferences.GetUnitPreference() == kImperial)
            tempUnits = UnitConverter.TempUnits.kFahrenheit;

        String unit = "";

            switch (tempUnits)
            {
                case kCelsius:
                    unit = "C";
                    break;
                case kFahrenheit:
                    unit = "F";
                    break;
            }

        if(mUnitConverter == null)
            return null;

        Integer value = mUnitConverter.GetTempInPreferredUnits(mForecastModel.mTemp, mForecastModel, tempUnits).intValue();

        return String.format(Locale.getDefault(),"%d", value) + "°" + unit;
    }



    @Override
    public String GetWindDescription()
    {
        if(mForecastModel.mWindSpd == null)
            return "Calm\n";

        UnitConverter.SpeedUnits speedUnits = UnitConverter.SpeedUnits.kMps;

        if(mPreferences == null || mPreferences.GetUnitPreference() == kImperial)
            speedUnits = UnitConverter.SpeedUnits.kMph;

        String unit = "";

        switch (speedUnits)
        {
            case kMps:
                unit = "mps";
                break;
            case kMph:
                unit = "mph";
                break;
        }

        if(mUnitConverter == null)
            return null;

        Integer value = mUnitConverter.GetSpeedInPreferredUnits(mForecastModel.mWindSpd, mForecastModel, speedUnits).intValue();

        return "Wind " + value.toString() + " at " + unit;
    }

    @Override
    public Observable GetObservable() {
        return this;
    }
}

File: app/src/main/java/org/mesonet/app/forecast/SemiDayForecastDataController.kt
package org.mesonet.app.forecast

import android.content.Context
import android.content.res.Configuration

import org.mesonet.app.R
import org.mesonet.app.dependencyinjection.PerFragment
import org.mesonet.app.formulas.DefaultUnits
import org.mesonet.app.formulas.UnitConverter
import org.mesonet.app.userdata.Preferences

import java.util.Locale
import java.util.Observable

import javax.inject.Inject

import org.mesonet.app.userdata.Preferences.UnitPreference.kImperial
import org.mesonet.app.userdata.Preferences.UnitPreference.kMetric


class SemiDayForecastDataController(internal var mContext: Context?, internal var mPreferences: Preferences?, internal var mUnitConverter: UnitConverter?, internal var mForecastModel: ForecastModel) : Observable(), ForecastData {

    fun SetData(inForecast: ForecastModel) {
        mForecastModel = inForecast

        setChanged()
        notifyObservers()
    }

    override fun GetTime(): String {
        var result = mForecastModel.mTime

        if (mContext != null && mContext!!.resources.getBoolean(R.bool.forceWrapForecasts) && !result!!.contains(" "))
            result += "\n"

        return result!!
    }

    override fun GetIconUrl(): String {
        return mForecastModel.mIconUrl!!.replace("http://www.nws.noaa.gov/weather/images/fcicons", "http://www.mesonet.org/images/fcicons-android").replace(".jpg", "@4x.png")
    }

    override fun GetStatus(): String {
        return mForecastModel.mStatus!! + "\n"
    }

    override fun GetHighOrLow(): String {
        return mForecastModel.mHighOrLow!!.name
    }

    override fun GetTemp(): String {
        var tempUnits: UnitConverter.TempUnits = UnitConverter.TempUnits.kCelsius

        if (mPreferences == null || mPreferences!!.GetUnitPreference() == kImperial)
            tempUnits = UnitConverter.TempUnits.kFahrenheit

        var unit = ""

        when (tempUnits) {
            UnitConverter.TempUnits.kCelsius -> unit = "C"
            UnitConverter.TempUnits.kFahrenheit -> unit = "F"
        }

        if (mUnitConverter == null)
            return ""

        val value = mUnitConverter!!.GetTempInPreferredUnits(mForecastModel.mTemp, mForecastModel, tempUnits)!!.toInt()

        return String.format(Locale.getDefault(), "%d", value) + "°" + unit
    }


    override fun GetWindDescription(): String {
        if (mForecastModel.mWindSpd == null)
            return "Calm\n"

        var speedUnits: UnitConverter.SpeedUnits = UnitConverter.SpeedUnits.kMps

        if (mPreferences == null || mPreferences!!.GetUnitPreference() == kImperial)
            speedUnits = UnitConverter.SpeedUnits.kMph

        var unit = ""

        when (speedUnits) {
            UnitConverter.SpeedUnits.kMps -> unit = "mps"
            UnitConverter.SpeedUnits.kMph -> unit = "mph"
        }

        if (mUnitConverter == null)
            return ""

        val value = mUnitConverter!!.GetSpeedInPreferredUnits(mForecastModel.mWindSpd, mForecastModel, speedUnits)!!.toInt()

        return "Wind " + value.toString() + " at " + unit
    }

    override fun GetObservable(): Observable {
        return this
    }
}
